### ISSUE TITLE
Results fluctuate heavily

### ISSUE THREAD

**Issue by yuehaowang** [graphdeeplearning/benchmarking-gnns#41]
Hi, thanks for creating this awesome project. However, I've encountered some problems on reproduction of results present in the paper and the leaderboard. Generally speaking, results are very unstable in my experiments.

### 1. Results fluctuate heavily.

On ZINC dataset, here are my results of GAT model :

- Trial 1: Mean test MAE = 0.38158. Under each seed (95, 41, 12, 35 and the same below): 0.3972, 0.3788, 0.3800, 0.3703
- Trial 2: Mean test MAE = 0.37115. Under each seed: 0.3636, 0.3785, 0.3769, 0.3656.

My results of GCN model:

- Trial 1: Mean test MAE = 0.37575. Under each seed:  0.3770, 0.3659, 0.3742, 0.3859
- Trial 2: Mean test MAE = 0.38070. Under each seed: 0.3946, 0.3668, 0.3764, 0.3850

I think we can not neglect these fluctuating errors since sometimes errors are around 0.01. And some recent SOTA models only outperform 0.01 over the best model in the benchmark paper.

It seems setting random seeds does not really work in this case. Is there any other requirements for reproduction? Do you think such fluctuation is tolerable?

### 2. I cannot reproduce results using checkpoints.

After I trained a GCN model on ZINC dataset, I reload the latest checkpoint and then try to reproduce test MAE through `evaluate_network_sparse` function. Unfortunately, test MAE from the checkpoint (0.3670) is different than test MAE from the result file (0.3757). Can we reproduce results via checkpoints? Could you please provide me some code of testing pipeline in case that I missed some important steps during reproduction?

Note: I changed `readout` to 'sum'. Besides, I didn't modify other hyper-parameters.
My experiments are conducted on Quadro P6000.

**Comment by chaitjo** [comment]
Hi @yuehaowang, thank you for your interest and for opening this issue; I hope we can improve the project together if indeed there is large fluctuation of results. I will try to reproduce your findings on my side and get back to you.

In the meantime, do let me know any additional details and changes you may have made, your hardware and Cuda setup, whether you followed our exact installation, etc.?

**Comment by yuehaowang** [comment]
- CUDA version: 10.1
- Hardware: Quadro P6000
- Configs:
```
params={'seed': 95, 'epochs': 1000, 'batch_size': 128, 'init_lr': 0.001, 'lr_reduce_factor': 0.5, 'lr_schedule_patience': 10, 'min_lr': 1e-05, 'weight_decay': 0.0, 'print_epoch_interval': 5, 'max_time': 12}

net_params={'L': 16, 'hidden_dim': 172, 'out_dim': 172, 'residual': True, 'readout': 'sum', 'in_feat_dropout': 0.0, 'dropout': 0.0, 'batch_norm': True, 'self_loop': False, 'device': device(type='cuda'), 'gpu_id': 1, 'batch_size': 128, 'max_order': 2, 'num_atom_type': 28, 'num_bond_type': 4, 'total_param': 505079}
```
- I run experiments in *screen* sessions instead of tmux.
- Packages are installed according to *environment_gpu.yml*

**Comment by chaitjo** [comment]
Thanks for the additional information, it is helpful as I diagnose this.

>CUDA version: 10.1
>...
>Packages are installed according to environment_gpu.yml

I am not sure if this is a big issue, but the env file will install DGL, PyTorch, etc. for CUDA 10.0 only, e.g. https://github.com/graphdeeplearning/benchmarking-gnns/blob/master/environment_gpu.yml#L19

Did you manually tweak the file to install for CUDA 10.1?

>I run experiments in screen sessions instead of tmux.

This should not be an issue in my experience.

**Comment by yuehaowang** [comment]
Thanks for your reply. I'd like to share my *environment_gpu.yml*:

```
name: benchmark_gnn
channels:
- pytorch 
- dglteam
- conda-forge
- fragcolor
- anaconda
- defaults
dependencies:
- cuda10.0
- cudatoolkit=10.0
- cudnn=7.6.5
- python=3.7.4
- python-dateutil=2.8.0
- pip=19.2.3
- pytorch=1.3
- torchvision==0.4.2
- pillow==6.1
- dgl-cuda10.0=0.4.2
- numpy=1.16.4
- matplotlib=3.1.0
- tensorboard=1.14.0
- tensorboardx=1.8
- future=0.18.2
- absl-py
- networkx=2.3
- scikit-learn=0.21.2
- scipy=1.3.0
- notebook=6.0.0
- h5py=2.9.0
- mkl=2019.4
- ipykernel=5.1.2
- ipython=7.7.0
- ipython_genutils=0.2.0
- ipywidgets=7.5.1
- jupyter=1.0.0
- jupyter_client=5.3.1
- jupyter_console=6.0.0
- jupyter_core=4.5.0
- plotly=4.1.1
- scikit-image=0.15.0
- requests==2.22.0
- tqdm==4.43.0
- pandas==1.1.0
- pip:
  - tensorflow-gpu==2.1.0
  - tensorflow-estimator==2.1.0
  - tensorboard==2.1.1
  - ogb==1.1.1
```

I think I didn't modify anything except I added *pandas* to the list of dependencies. And another info about my CUDA:

- Driver Version: 430.50

**Comment by chaitjo** [comment]
Thanks; for future experiments with this codebase, I would suggest either downgrading CUDA 10.1-->10.0, or upgrading those GPU dependent packages to use CUDA 10.1 (e.g. DGL, PyTorch, Cuda toolkit, etc.) as the incompatibility between Python packages' expected CUDA version and the version on your system can sometimes cause issues, too.

**Comment by chaitjo** [comment]
Hi @yuehaowang, I tried running the same seeds on ZINC multiple times and I was able to reproduce my results. I believe your issue is due to differences between the library versions/compatibility and CUDA version on your hardware. Basically, you should either downgrade system CUDA to 10.0 or upgrade DGL/PyTorch/core GPU packages to be compatible with latest CUDA.
